hudi-cli/src/main/scala/org/apache/hudi/cli/DedupeSparkJob.scala

```diff
@@ -103,24 +105,51 @@ class DedupeSparkJob(basePath: String,
     // Mark all files except the one with latest commits for deletion
     dupeMap.foreach(rt => {
       val (key, rows) = rt
-      var maxCommit = -1L
-
-      rows.foreach(r => {
-        val c = r(3).asInstanceOf[String].toLong
-        if (c > maxCommit)
-          maxCommit = c
-      })
-
-      rows.foreach(r => {
-        val c = r(3).asInstanceOf[String].toLong
-        if (c != maxCommit) {
-          val f = r(2).asInstanceOf[String].split("_")(0)
-          if (!fileToDeleteKeyMap.contains(f)) {
-            fileToDeleteKeyMap(f) = HashSet[String]()
+
+      if (useCommitTimeForDedupe) {
+        /*
+        This corresponds to the case where duplicates got created due to INSERT and have never been updated.
+         */
+        var maxCommit = -1L
+
+        rows.foreach(r => {
+          val c = r(3).asInstanceOf[String].toLong
+          if (c > maxCommit)
+            maxCommit = c
+        })
+        rows.foreach(r => {
+          val c = r(3).asInstanceOf[String].toLong
+          if (c != maxCommit) {
+            val f = r(2).asInstanceOf[String].split("_")(0)
+            if (!fileToDeleteKeyMap.contains(f)) {
+              fileToDeleteKeyMap(f) = HashSet[String]()
+            }
+            fileToDeleteKeyMap(f).add(key)
           }
-          fileToDeleteKeyMap(f).add(key)
+        })
+      } else {
+        /*
+        This corresponds to the case where duplicates have been updated at least once.
+        Once updated, duplicates are bound to have same commit time unless forcefully modified.
+         */
+        val size = rows.size - 1
+        var i = 0
+        val loop = new Breaks
+        loop.breakable {
```

Comment:
It's better not use break here, `rows.init` also can get the rows will be delete.

Reply:
Right. Thank you for suggesting this, I am not hands on at scala properly. :)